README.md
# opening_hours.js

[![Build Status](https://travis-ci.org/opening-hours/opening_hours.js.svg?branch=master)](https://travis-ci.org/opening-hours/opening_hours.js)
[![license NPM](https://img.shields.io/npm/l/opening_hours.svg)][ohlib.github]
[![required Node.js version](https://img.shields.io/node/v/opening_hours.svg)][ohlib.github]
[![version NPM](https://img.shields.io/npm/v/opening_hours.svg)][ohlib.npmjs]
[![monthly downloads via NPM](https://img.shields.io/npm/dm/opening_hours.svg)][ohlib.npmjs]
[![total downloads via NPM](https://img.shields.io/npm/dt/opening_hours.svg)][ohlib.npmjs]

<!-- https://img.shields.io/github/license/opening-hours/opening_hours.js.svg -->
<!-- https://img.shields.io/github/downloads/opening-hours/opening_hours.js/total.svg -->

## Summary

The [opening_hours][Key:opening_hours] tag is used in [OpenStreetMap](https://openstreetmap.org) to describe time ranges when a facility (for example, a café) is open. As it has pretty [complex syntax][oh:specification] which requires special parsing and additional processing to extract some useful information (e.g. whether a facility is open at a specific time, next time it's going to open/close, or a readable set of working hours), this library is being developed.

Examples of some complex opening_hours values:

```
Mo,Tu,Th,Fr 12:00-18:00; Sa,PH 12:00-17:00; Th[3],Th[-1] off
Mo-Fr 12:00-18:00; We off; Sa,PH 12:00-17:00; Th[3],Th[-1] off
```

a library which is open from 12:00 to 18:00 on workdays (Mo-Fr) except Wednesday, and from 12:00 to 17:00 on Saturday and public holidays. It also has breaks on the third and last Thursday of each month.

```
open; Tu-Su 08:30-09:00 off; Tu-Su,PH 14:00-14:30 off; Mo 08:00-13:00 off
```

around-the-clock shop with some breaks.

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [evaluation tool](#evaluation-tool)
- [Install](#install)
  - [For Developer](#for-developer)
  - [Web developer](#web-developer)
  - [NodeJS developer](#nodejs-developer)
- [Versions](#versions)
- [Synopsis](#synopsis)
- [Library API](#library-api)
  - [Simple API](#simple-api)
  - [High-level API](#high-level-api)
  - [Iterator API](#iterator-api)
- [Features](#features)
  - [Time ranges](#time-ranges)
  - [Points in time](#points-in-time)
  - [Weekday ranges](#weekday-ranges)
  - [Holidays](#holidays)
  - [Month ranges](#month-ranges)
  - [Monthday ranges](#monthday-ranges)
  - [Week ranges](#week-ranges)
  - [Year ranges](#year-ranges)
  - [States](#states)
  - [Comments](#comments)
- [Testing](#testing)
  - [Regression testing](#regression-testing)
  - [Testing with real data](#testing-with-real-data)
    - [Large scale](#large-scale)
    - [Small scale](#small-scale)
  - [Test it yourself (the geeky way)](#test-it-yourself-the-geeky-way)
- [Performance](#performance)
- [Used by other projects](#used-by-other-projects)
  - [YoHours](#yohours)
- [Bindings and ports](#bindings-and-ports)
- [Other implementations](#other-implementations)
- [Related links](#related-links)
- [ToDo](#todo)
- [How to contribute](#how-to-contribute)
  - [Translating the evaluation tool and the map](#translating-the-evaluation-tool-and-the-map)
  - [Translating error messages and warnings](#translating-error-messages-and-warnings)
  - [Holidays](#holidays-1)
  - [Core code](#core-code)
    - [Commit hooks](#commit-hooks)
    - [Documentation](#documentation)
- [Authors](#authors)
- [Contributors](#contributors)
- [Credits](#credits)
- [Stats](#stats)
- [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## evaluation tool

Please have a look at the [evaluation tool][] which can give you an impression how this library can be used and what it is capable of.

A mirror is setup up under: https://openingh.ypid.de/evaluation_tool/


## Install

### For Developer
Just clone the repository:

```Shell
git clone --recursive https://github.com/opening-hours/opening_hours.js.git
```

and install it’s dependencies (execute inside the repository):
```Shell
npm install
```

### Web developer

If you are a web developer and want to use this library you can do so by including the current version from here:

```
http://openingh.openstreetmap.de/opening_hours.js/opening_hours+deps.min.js
```

To get started checkout the [simple_index.html](/examples/simple_index.html) file.

### NodeJS developer

This library is packaged with npm and is available under the name [opening_hours][ohlib.npmjs] so you should have no problems using it.


## Versions

The version number consists of a major release, minor release and patch level (separated by a dot).

For version 2.2.0 and all later, [Semantic Versioning](http://semver.org/spec/v2.0.0.html) is used:

* The major release is only increased if the release breaks backward compatibility.
* The minor release is increased if new features are added.
* The patch level is increased to bundle a bunch of commits (minor changes like bug fixes and improvements) into a new tested version.

Check [releases on GitHub] for a list of the releases and their Changelog.

## Synopsis

```JavaScript
var oh = new opening_hours('We 12:00-14:00');

var from = new Date("01 Jan 2012");
var to   = new Date("01 Feb 2012");

// high-level API
{
    var intervals = oh.getOpenIntervals(from, to);
    for (var i in intervals)
        console.log('We are ' + (intervals[i][2] ? 'maybe ' : '')
            + 'open from ' + (intervals[i][3] ? '("' + intervals[i][3] + '") ' : '')
            + intervals[i][0] + ' till ' + intervals[i][1] + '.');

    var duration_hours = oh.getOpenDuration(from, to).map(function(x) { return x / 1000 / 60 / 60 });
    if (duration_hours[0])
        console.log('For the given range, we are open for ' + duration_hours[0] + ' hours');
    if (duration_hours[1])
        console.log('For the given range, we are maybe open for ' + duration_hours[1] + ' hours');
}

// helper function
function getReadableState(startString, endString, oh, past) {
    if (past === true) past = 'd';
    else past = '';

    var output = '';
    if (oh.getUnknown()) {
        output += ' maybe open'
            + (oh.getComment() ? ' but that depends on: "' + oh.getComment() + '"' : '');
    } else {
        output += ' ' + (oh.getState() ? 'open' : 'close' + past)
            + (oh.getComment() ? ', comment "' + oh.getComment() + '"' : '');
    }
    return startString + output + endString + '.';
}

// simple API
{
    var state      = oh.getState(); // we use current date
    var unknown    = oh.getUnknown();
    var comment    = oh.getComment();
    var nextchange = oh.getNextChange();

    console.log(getReadableState('We\'re', '', oh, true));

    if (typeof nextchange === 'undefined')
        console.log('And we will never ' + (state ? 'close' : 'open'));
    else
        console.log('And we will '
            + (oh.getUnknown(nextchange) ? 'maybe ' : '')
            + (state ? 'close' : 'open') + ' on ' + nextchange);
}

// iterator API
{
    var iterator = oh.getIterator(from);

    console.log(getReadableState('Initially, we\'re', '', iterator, true));

    while (iterator.advance(to))
        console.log(getReadableState('Then we', ' at ' + iterator.getDate(), iterator));

    console.log(getReadableState('And till the end we\'re', '', iterator, true));
}
```

## Library API

*   `var oh = new opening_hours('We 12:00-14:00', nominatim_object, mode);`

    *   value (mandatory, type: string): Constructs opening_hours object, given the opening_hours tag value.

        Throws an error string if the expression is malformed or unsupported.

    *   nominatim_object (optional, type: object or null): In order to calculate the correct times for variable times (e.g. sunrise, dusk, see under [Time ranges][ohlib.time-ranges]) the coordinates are needed. To apply the correct holidays (PH) and school holidays (SH) the country code and the state is needed. The only thing you as programmer need to know are the coordinates or preferably the OSM id (for the node, way or relation) of the facility (where the opening hours do apply) anything else can be queried for using [reverse geocoding with Nominatim][Nominatim]. So just use as second parameter the returned JSON from [Nominatim] (example URL: https://nominatim.openstreetmap.org/reverse?format=json&lat=49.5487429714954&lon=9.81602098644987&zoom=5&addressdetails=1) and you are good to go. Note that this second parameter is optional. The data returned by Nominatim should be in the local language (the language of the country for which the opening hours apply). If not, *accept-language* can be used as parameter in the request URL.

        The `nominatim_object` can also be `null` in which case a default location will be used.
        This can be used if you don’t care about correct opening hours for more complex opening_hours values.

    *   optional_conf_parm (optional, either of type number of object):

        If this parameter is of the type number then it is interpreted as 'mode' (see below). For the type object, the following keys are defined.

        *   'mode' (type: (integer) number, default: 0): In OSM, the syntax originally designed to describe opening hours, is now used to describe a few other things as well. Some of those other tags work with points in time instead of time ranges. To support this the mode can be specified. *Note that it is recommended to not use the mode parameter directly but instead use the tag_key parameter.* If there is no mode specified, opening_hours.js will only operate with time ranges and will throw an error when points in times are used in the value.

            * 0: time ranges (opening_hours, lit, …) default
            * 1: points in time
            * 2: both (time ranges and points in time, used by collection_times, service_times, …)


CHANGELOG.rst
Changelog
=========

All notable changes to opening_hours.js will be documented in this file.

This project adheres to `Semantic Versioning <http://semver.org/spec/v2.0.0.html>`__
and `human-readable changelog <http://keepachangelog.com/en/0.3.0/>`__.

Note that most of the v2.X.Z releases have not been added to the changelog yet.

Authors
-------

* [AMDmi3_] - Dmitry Marakasov (original author)
* [ypid_] - Robin Schneider (author, maintainer)

.. _AMDmi3: https://github.com/AMDmi3
.. _ypid: https://me.ypid.de/

Contributors
------------

* [putnik_] - Sergey Leschina
* [Cactusbone_] - Charly Koza
* [don-vip_] - Vincent Privat
* [sesam_] - Simon B.
* [NonnEmilia_]
* [damjang_]
* [jgpacker_] - João G. Packer
* [openfirmware_] - James Badger
* [burrbull_] - Zgarbul Andrey
* [blorger_] - Blaž Lorger
* [dmromanov_] - Dmitrii Romanov
* [maxerickson_]
* [amenk_] - Alexander Menk
* [edqd_]
* [simon04_] - Simon Legner
* [MKnight_] - Michael Knight
* [elgaard_] - Niels Elgaard Larsen
* [afita_] - Adrian Fita
* [sanderd17_]
* [marcgemis_]
* [drMerry_]
* [endro_]
* [rmikke_] - Ryszard Mikke
* [VorpalBlade_] - Arvid Norlander
* [mmn_] - Mikael Nordfeldth
* [adrianojbr_]
* [AndreasTUHU_]
* [mkyral_] - Marián Kyral
* [pke_] - Philipp Kursawe
* [spawn-guy_] - Paul Rysiavets
* [bugvillage_]
* [ItsNotYou_]
* [chiak597_]
* [skifans_] - Alex
* [shouze_] - Sébastien HOUZÉ
* [Gazer75_]
* [yourock17_]
* [urbalazs_] - Balázs Úr
* [HolgerJeromin_] - Holger Jeromin
* [Gerw88_]
* [sommerluk_] - Lukas Sommer
* [Discostu36_] - Michael
* [debyos_] - Gábor Babos
* [StephanGeorg_] - Stephan Georg

Thanks very much to all contributors!

.. _putnik: https://github.com/putnik
.. _Cactusbone: https://github.com/Cactusbone
.. _don-vip: https://github.com/don-vip
.. _sesam: https://github.com/sesam
.. _NonnEmilia: https://github.com/NonnEmilia
.. _damjang: https://github.com/damjang
.. _jgpacker: https://github.com/jgpacker
.. _openfirmware: https://github.com/openfirmware
.. _burrbull: https://github.com/burrbull
.. _blorger: https://github.com/blorger
.. _dmromanov: https://github.com/dmromanov
.. _maxerickson: https://github.com/maxerickson
.. _amenk: https://github.com/amenk
.. _edqd: https://github.com/edqd
.. _simon04: https://github.com/simon04
.. _MKnight: https://github.com/dex2000
.. _elgaard: https://github.com/elgaard
.. _afita: https://github.com/afita
.. _sanderd17: https://github.com/sanderd17
.. _marcgemis: https://github.com/marcgemis
.. _drMerry: https://github.com/drMerry
.. _endro: https://github.com/endro
.. _rmikke: https://github.com/rmikke
.. _VorpalBlade: https://github.com/VorpalBlade
.. _mmn: https://blog.mmn-o.se/
.. _adrianojbr: https://github.com/adrianojbr
.. _AndreasTUHU: https://github.com/AndreasTUHU
.. _mkyral: https://github.com/mkyral
.. _pke: https://github.com/pke
.. _bugvillage: https://github.com/bugvillage
.. _ItsNotYou: https://github.com/ItsNotYou
.. _spawn-guy: https://github.com/spawn-guy
.. _chiak597: https://github.com/chiak597
.. _skifans: https://github.com/skifans
.. _shouze: https://github.com/shouze
.. _Gazer75: https://github.com/Gazer75
.. _yourock17: https://github.com/yourock17
.. _urbalazs: https://github.com/urbalazs
.. _HolgerJeromin: https://github.com/HolgerJeromin
.. _Gerw88: https://github.com/Gerw88
.. _sommerluk: https://github.com/sommerluk
.. _Discostu36: https://github.com/Discostu36
.. _debyos: https://github.com/debyos
.. _StephanGeorg: https://github.com/StephanGeorg

Supporters
~~~~~~~~~~

* `iMi digital`_
* AddisMap_

Thanks for helping by allowing employees to work on the project during work hours!

.. _iMi digital: http://www.imi-digital.de/
.. _AddisMap: https://www.addismap.com/

master_ - unreleased
--------------------

.. _master: https://github.com/opening-hours/opening_hours.js/compare/v3.5.0...master

`v3.6.0 milestone <https://github.com/opening-hours/opening_hours.js/issues?q=milestone%3Av3.6.0+is%3Aclosed>`__

Added
~~~~~

* Public holiday definitions added:

  * England and Wales [skifans_, simon04_]
  * Norway [Gazer75_]
  * Australia [yourock17_, ypid_]
  * Ireland [Gerw88_, ypid_]
  * Ivory Coast [sommerluk_, ypid_]

* Added Easter Sunday to Slovak holidays. [chiak597_]
* Localizations added: Hungarian [urbalazs_, debyos_]

* School holiday definitions added:

  * Germany 2017 until 2024 [ypid_]
  * Austria [simon04_]

* Translate error tolerance warnings into German. [ypid_]
* Add additional warnings:

  * Misused ``.`` character. Example: ``Jan 01,Dec 24.-25.``. [ypid_]
  * Trailing ``,`` after time selector. Example: ``We 12:00-18:00,``. [ypid_]
  * Additional rule which evaluates to closed. Example: ``Mo-Fr 10:00-20:00, We off``. [ypid_]
  * Value consists of multiple rules each only using a time selector. Example: ``11:30-14:30;17:30-23:00``. [ypid]
  * Potentially missing use of ``<additional_rule_separator>`` if the previous
    rule has a time range which wraps over midnight and the current rule

js/helpers.js
/* Constants {{{ */
var nominatim_api_url = 'https://nominatim.openstreetmap.org/reverse';
// var nominatim_api_url = 'https://open.mapquestapi.com/nominatim/v1/reverse.php';

var evaluation_tool_colors = {
    'ok': '#ADFF2F',
    'warn': '#FFA500',
    'error': '#DEB887',
};
/* }}} */

// load nominatim_data in JOSM {{{
// Using a different way to load stuff in JOSM than https://github.com/rurseekatze/OpenLinkMap/blob/master/js/small.js
// prevent josm remote plugin of showing message
// FIXME: Warning in console. Encoding stuff.
function josm(url_param) {
    var xhr = new XMLHttpRequest();
    xhr.open('GET', 'https://localhost:8111/' + url_param, true);      // true makes this call asynchronous
    xhr.onreadystatechange = function () {    // need eventhandler since our call is async
        if ( xhr.status !== 200 ) {
            alert(i18n.t('texts.JOSM remote conn error'));
        }
    };
    xhr.send(null);
}
// }}}

// add calculation for calendar week to date {{{
function dateAtWeek(date, week) {
    var minutes_in_day = 60 * 24;
    var msec_in_day    = 1000 * 60 * minutes_in_day;
    var msec_in_week   = msec_in_day * 7;

    var tmpdate = new Date(date.getFullYear(), 0, 1);
    tmpdate.setDate(1 - (tmpdate.getDay() + 6) % 7 + week * 7); // start of week n where week starts on Monday
    return Math.floor((date - tmpdate) / msec_in_week);
}
// }}}

/*
 * The names of countries and states are localized in OSM and opening_hours.js
 * (holidays) so we need to get the localized names from Nominatim as well.
 */
function reverseGeocodeLocation(query, guessed_language_for_location, on_success, on_error) {
    if (typeof on_error === 'undefined') {
        on_error = function() { };
    }

    if (query === '&lat=48.7769&lon=9.1844') {
        /* Cached response to avoid two queries for each usage of the tool. */
        return on_success({"place_id":"159221147","licence":"Data © OpenStreetMap contributors, ODbL 1.0. https:\/\/www.openstreetmap.org\/copyright","osm_type":"relation","osm_id":"62611","lat":"48.6296972","lon":"9.1949534","display_name":"Baden-Württemberg, Deutschland","address":{"state":"Baden-Württemberg","country":"Deutschland","country_code":"de"},"boundingbox":["47.5324787","49.7912941","7.5117461","10.4955731"]});
    }

    var nominatim_api_url_template_query = nominatim_api_url
        + '?format=json'
        + query
        + '&zoom=5'
        + '&addressdetails=1'
        + '&email=ypid23@aol.de';

    var nominatim_api_url_query = nominatim_api_url_template_query;
    if (typeof accept_lanaguage === 'string') {
        nominatim_api_url_query += '&accept-language=' + guessed_language_for_location;
    }

    $.getJSON(nominatim_api_url_query, function(nominatim_data) {
        // console.log(JSON.stringify(nominatim_data, null, '\t'));
        if (nominatim_data.address.country_code === guessed_language_for_location) {
            on_success(nominatim_data);
        } else {
            nominatim_api_url_query += '&accept-language=' + mapCountryToLanguage(nominatim_data.address.country_code);
            $.getJSON(nominatim_api_url_query, function(nominatim_data) {
                on_success(nominatim_data);
            }).error(on_error);
        }
    }).error(on_error);
}

function submitenter(myfield,e) {
    Evaluate();
    // var keycode;
    // if (window.event) keycode = window.event.keyCode;
    // else if (e) keycode = e.which;
    // else return true;

    // if (keycode === 13) {
    //     Evaluate();
    //     return false;
    // } else
    //     return true;
}

/* JS for toggling examples on and off {{{ */
function toggle(control){
    var elem = document.getElementById(control);

    if (elem.style.display === "none") {
        elem.style.display = "block";
    } else {
        elem.style.display = "none";
    }
}
/* }}} */

function copyToClipboard(text) {
    window.prompt('Copy to clipboard: Ctrl+C, Enter', text);
}

var lat, lon, string_lat, string_lon, nominatim;
var date;

function Evaluate (offset, reset) {
    if (typeof offset === 'undefined') {
        offset = 0;
    }

    if (document.forms.check.elements['lat'].value !== string_lat || document.forms.check.elements['lon'].value !== string_lon) {
        string_lat = document.forms.check.elements['lat'].value;
        string_lon = document.forms.check.elements['lon'].value;
        lat = parseFloat(string_lat);
        lon = parseFloat(string_lon);
        if (typeof lat !== 'number' || typeof lon !== 'number') {
            if (typeof lat !== 'number') {
                document.forms.check.elements['lat'].value = default_lat;
            }
            if (typeof lon !== 'number') {
                document.forms.check.elements['lon'].value = default_lon;
            }
            console.log('Please enter numbers for latitude and longitude.');
            return;
        }
        reverseGeocodeLocation(
            '&lat=' + lat + '&lon=' + lon,
            mapCountryToLanguage(i18n.lng()),
            function(nominatim_data) {
                nominatim = nominatim_data;
                document.forms.check.elements['cc'].value    = nominatim.address.country_code;
                document.forms.check.elements['state'].value = nominatim.address.state;
                Evaluate();
            },
            function() {
                /* Set fallback Nominatim answer to allow using the evaluation tool even without Nominatim. */
                alert("Reverse geocoding of the coordinates using Nominatim was not successful. The evaluation of features of the opening_hours specification which depend this information will be unreliable. Otherwise, this tool will work as expected using a fallback answer. You might want to check your browser settings to fix this.");
                nominatim = {"place_id":"44651229","licence":"Data \u00a9 OpenStreetMap contributors, ODbL 1.0. https:\/\/www.openstreetmap.org\/copyright","osm_type":"way","osm_id":"36248375","lat":"49.5400039","lon":"9.7937133","display_name":"K 2847, Lauda-K\u00f6nigshofen, Main-Tauber-Kreis, Regierungsbezirk Stuttgart, Baden-W\u00fcrttemberg, Germany, European Union","address":{"road":"K 2847","city":"Lauda-K\u00f6nigshofen","county":"Main-Tauber-Kreis","state_district":"Regierungsbezirk Stuttgart","state":"Baden-W\u00fcrttemberg","country":"Germany","country_code":"de","continent":"European Union"}};
                document.forms.check.elements['cc'].value    = nominatim.address.country_code;
                document.forms.check.elements['state'].value = nominatim.address.state;
                Evaluate();
            }
        );
    }

    date = reset
        ? new Date()
        : new Date(
            document.forms.check.elements['yyyy'].value,
            document.forms.check.elements['mm'].selectedIndex,
            document.forms.check.elements['dd'].value,
            document.forms.check.elements['HH'].value,
            parseInt(document.forms.check.elements['MM'].value),
            offset
        );

    function u2 (v) { return v>=0 && v<10 ? "0"+v : v; }

    document.forms.check.elements['yyyy'].value       = date.getFullYear();
    document.forms.check.elements['mm'].selectedIndex = date.getMonth();
    document.forms.check.elements['dd'].value         = u2(date.getDate());
    document.forms.check.elements['HH'].value         = u2(date.getHours());
    document.forms.check.elements['MM'].value         = u2(date.getMinutes());
    document.forms.check.elements['wday'].value       = moment.weekdays(date.getDay());
    document.forms.check.elements['week'].value       = 'W'+u2(dateAtWeek(date, 0) + 1);

    var show_time_table         = document.getElementById('show_time_table');
    var show_warnings_or_errors = document.getElementById('show_warnings_or_errors');
    var show_results            = document.getElementById('show_results');

    show_warnings_or_errors.innerHTML = '';

    var crashed = false;
    var value = document.forms.check.elements['expression'].value;
    var diff_value = document.forms.check.elements['diff_value'].value;
    var mode = parseInt(document.getElementById('mode').selectedIndex);
    try {
        var oh = new opening_hours(value, nominatim, {
            'mode': mode,
            'warnings_severity': 7,
            'locale': i18n.lng()
        });
        var it = oh.getIterator(date);
    } catch (err) {
        crashed = err;
        show_warnings_or_errors.innerHTML = '<p>' + i18n.t('texts.filter.error') + ':<br />'
            + '<textarea rows="' + crashed.split('\n').length + 1 + '" style="width: 100%" name="WarnErrors" readonly="readonly">' + crashed
            + '</textarea></p>';
        show_time_table.innerHTML = '';
        show_results.innerHTML    = '';
    }

    show_time_table.innerHTML = '<a href="javascript:josm(\'import?url=' + encodeURIComponent('https://overpass-api.de/api/xapi_meta?*[opening_hours='
        + document.forms.check.elements['expression'].value + ']') + '\')">' + i18n.t('texts.load all with JOSM') + '</a><br />';
    if (!crashed) {
        var prettified = oh.prettifyValue({});
        var prettified_value_array = oh.prettifyValue({
            // conf: { locale: i18n.lng() },
            get_internals: true,
        });
        // var prettified_newline_sep = oh.prettifyValue({ conf: { locale: i18n.lng(), rule_sep_string: '\n', print_semicolon: false } });
        show_results.innerHTML = '<p><span class="hd">' + i18n.t('words.status') + ':</span>'
            + '<input class="nostyle" size="10" name="status" readonly="readonly" />'
            + '<input class="nostyle" size="60" name="comment" readonly="readonly" />'
            + '</p>' + '<p><span class="hd">'
            + i18n.t('texts.MatchingRule') + ':</span>'
            + '<input class="nostyle w100" name="MatchingRule" readonly="readonly" />'
            + '</p>';
        var used_selectors = { };
        var value_explanation =
            i18n.t('texts.prettified value for displaying') + ':<br />'
            + '<p class="value_explanation">';
        // console.log(JSON.stringify(prettified_value_array, null, '    '));
        // console.log(JSON.stringify(prettified_value_array, null, '    '));
        for (var nrule = 0; nrule < prettified_value_array[0].length; nrule++) {
            if (nrule !== 0) {
                var rule_separator = (
                    prettified_value_array[1][nrule][1]
                        ? ' ||'
                        : (
                            prettified_value_array[1][nrule][0][0][1] === 'rule separator'
                            ? ','
                            : ';'
                        )
                );
                value_explanation +=
                    '<span title="'
                    + i18n.t('texts.rule separator ' + rule_separator) + '"'
                    + ' class="rule_separator"><a target="_blank" class="specification" href="'
                    + specification_url + '#section:rule_separators'
                    + '">' + rule_separator + '</a></span><br>';
            }
            value_explanation += '<span class="one_rule">';
            for (var nselector = 0, sl = prettified_value_array[0][nrule].length; nselector < sl; nselector++) {
                var selector_type  = prettified_value_array[0][nrule][nselector][0][2];
                var selector_value = prettified_value_array[0][nrule][nselector][1];
                var fragment_identifier;
                switch(selector_type) {
                    case '24/7':
                        fragment_identifier = 'selector_sequence';
                        break;
                    case 'state':
                        fragment_identifier = 'section:rule_modifier';
                        break;
                    case 'comment':
                        fragment_identifier = 'comment';
                        break;
                    default:
                        fragment_identifier = 'selector:' + selector_type;
                }
                value_explanation += '<span title="'
                    + i18n.t('words.' + (selector_type.match(/(?:state|comment)/) ? 'modifier' : 'selector'), { name: selector_type }) + '"'
                    + ' class="' + selector_type + '"><a target="_blank" class="specification" href="'
                    + specification_url + '#' + fragment_identifier
                    + '">' + selector_value + '</a></span>';
                if (nselector + 1 < sl)
                    value_explanation += ' ';
                used_selectors[selector_type] = true;
            }
            // console.log(value_explanation);
            value_explanation += '</span>';
        }
        value_explanation += '</p></div>';
        if (YoHoursChecker.canRead(value)) {
            value_explanation = i18n.t('texts.refer to yohours', { href: 'http://github.pavie.info/yohours/?oh=' + value })
            + '<br>'
            + value_explanation;
        }

        if (diff_value.length > 0) {
          var is_equal_to;
          try {
              is_equal_to = oh.isEqualTo(new opening_hours(diff_value, nominatim, {
                  'mode': mode,
                  'warnings_severity': 7,
                  'locale': i18n.lng()
              }));
          } catch (err) {
              $('input#diff_value').css({'background-color' : evaluation_tool_colors.error})
          }
          if (typeof is_equal_to === 'object') {
            if (is_equal_to[0]) {
              $('input#diff_value').css({'background-color' : evaluation_tool_colors.ok})
            } else {
              $('input#diff_value').css({'background-color' : evaluation_tool_colors.warn})
              var human_readable_not_equal_output = jQuery.extend(true, {}, is_equal_to[1])
              if (typeof human_readable_not_equal_output.deviation_for_time === 'object') {
                human_readable_not_equal_output.deviation_for_time = {};
                for (var time_code in is_equal_to[1].deviation_for_time) {
                  console.log(time_code);
                  var time_string = new Date(parseInt(time_code)).toLocaleString();
                  human_readable_not_equal_output.deviation_for_time[time_string] =
                    is_equal_to[1].deviation_for_time[time_code];
                }
              }
              value_explanation = JSON.stringify(human_readable_not_equal_output, null, '    ')
                + '<br>'
                + value_explanation;
            }
          }
        }

        show_warnings_or_errors.innerHTML = value_explanation;

        // if (prettified_newline_sep.split('\n').length > 1)
            // show_results.innerHTML += '<p>' + i18n.t('texts.prettified value for displaying') + ':<br />'
                // + '<textarea rows="' + prettified_newline_sep.split('\n').length
                // + '" style="width: 100%" name="prettifiedValueNewlineSep" readonly="readonly">'
                // + prettified_newline_sep + '</textarea></p>';


        document.forms.check.elements['comment'].value = typeof it.getComment() !== 'undefined'
            ? it.getComment() : i18n.t('words.no') + ' ' + i18n.t('words.comment');
        document.forms.check.elements['status'].value = (it.getState() ? i18n.t('words.open')
            : (it.getUnknown() ? i18n.t('words.unknown') : i18n.t('words.closed')));
        var rule_index    = it.getMatchingRule();
        document.forms.check.elements['MatchingRule'].value = typeof rule_index === 'undefined'
            ? i18n.t('words.none') : oh.prettifyValue({ 'rule_index': rule_index });

        if (prettified !== value) {
            show_warnings_or_errors.innerHTML = '<p>' + i18n.t('texts.prettified value',
                { copyFunc: 'javascript:newValue(\'' + prettified.replace(/"/g, '&quot;') + '\')' }) + ':<br />'
                + '<input style="width: 100%" onclick="javascript:newValue(\'' + prettified.replace(/"/g, '&quot;') + '\')" id="prettifiedValue" name="prettifiedValue" value="' + prettified.replace(/"/g, '&quot;') + '" /></p>';
        }

        var warnings = oh.getWarnings();
        if (warnings.length > 0) {
            show_warnings_or_errors.innerHTML += '<p>' + i18n.t('texts.filter.error') + ':<br />'
                + '<textarea rows="' + (warnings.length + 1)
                + '" style="width: 100%" name="WarnErrors" readonly="readonly">' + warnings.join('\n')
                + '</textarea></p>';
        }

        if (prettified.length > 255) {
            show_warnings_or_errors.innerHTML += '<p>' + i18n.t('texts.value to long for osm',
                { pretLength: prettified.length, valLength: value.length, maxLength: 255 }) + '</p>';
        }

        show_time_table.innerHTML += OpeningHoursTable.drawTableAndComments(oh, it);
    }
}

function EX (element) {
    newValue(element.innerHTML);
    return false;
}

function newValue(value) {
    if (typeof document.forms.check.elements['prettifiedValue'] === 'object') {
        document.forms.check.elements['prettifiedValue'].focus();
        document.forms.check.elements['prettifiedValue'].focus();
    }
    document.forms.check.elements['expression'].value = value;
    Evaluate();
}

function permalink () {
    var exp = document.getElementById('expression').value;
    var diff_value = document.getElementById('diff_value').value;
    var lat = document.getElementById('lat').value;
    var lon = document.getElementById('lon').value;
    var mode = document.getElementById('mode').selectedIndex;

    var permalink_url_query='?EXP='+encodeURIComponent(exp)+'&lat='+lat+'&lon='+lon+'&mode='+mode;

    if (document.getElementById('permalink-include-timestamp').checked) {
        permalink_url_query += '&DATE='+date.getTime();
    }
    if (diff_value !== '') {
        permalink_url_query += '&diff_value='+encodeURIComponent(diff_value);
    }

    location = location.protocol+'//'+location.host+location.pathname+permalink_url_query;
}

function setCurrentPosition() {
    if(navigator.geolocation) {
        navigator.geolocation.getCurrentPosition(onPositionUpdate);
    }
}

function onPositionUpdate(position) {
    var lat = position.coords.latitude;
    var lng = position.coords.longitude;
    document.getElementById('lat').value = lat;
    document.getElementById('lon').value = lng;
    Evaluate();
    console.log("Current position: " + lat + " " + lng);
}
window.onload = function () {
    var prmarr = window.location.search.replace( "?", "" ).split("&");
    var params = {};
    var customCoords = false;

    for ( var i = 0; i < prmarr.length; i++) {
        var tmparr = prmarr[i].split("=");
        params[tmparr[0]] = tmparr[1];
    }
    if (typeof params['EXP'] !== 'undefined') {
        document.forms.check.elements['expression'].value = decodeURIComponent(params['EXP']);
    }
    if (typeof params['diff_value'] !== 'undefined') {
        document.forms.check.elements['diff_value'].value = decodeURIComponent(params['diff_value']);
    }
    if (typeof params['lat'] !== 'undefined') {
        document.forms.check.elements['lat'].value = decodeURIComponent(params['lat']);
        customCoords = true;
    }
    if (typeof params['lon'] !== 'undefined') {
        document.forms.check.elements['lon'].value = decodeURIComponent(params['lon']);
        customCoords = true;
    }
    if (typeof params['mode'] !== 'undefined') {
        document.forms.check.elements['mode'].value = decodeURIComponent(params['mode']);
    }
    if (typeof params['DATE'] !== 'undefined') {
        var crashed = true;
        try {
            date = new Date(parseInt(params['DATE']));
            crashed = false;
        } catch (err) {
            console.error(err);
        }
        if (!crashed) {
            document.forms.check.elements['yyyy'].value       = date.getFullYear();
            document.forms.check.elements['mm'].selectedIndex = date.getMonth();
            document.forms.check.elements['dd'].value         = date.getDate();
            document.forms.check.elements['HH'].value         = date.getHours();
            document.forms.check.elements['MM'].value         = date.getMinutes();
        }
        Evaluate(0, false);
    } else {
        Evaluate(0, true);
    }
    if (navigator.geolocation && !customCoords) {
        navigator.geolocation.getCurrentPosition(onPositionUpdate);
    };
};
/* }}} */

$(document).ready(function () {
    var permalink = document.getElementById('permalink');
    if (permalink) {
        var checkbox = document.createElement('input');
        checkbox.type = 'checkbox';
        checkbox.name = 'name';
        checkbox.value = 'value';
        checkbox.id = 'permalink-include-timestamp';
        checkbox.checked = true;

        var label = document.createElement('label')
        label.htmlFor = 'permalink-include-timestamp';
        label.appendChild(document.createTextNode(i18n.t('texts.include timestamp?')));

        permalink.appendChild(label);
        permalink.appendChild(checkbox);
    }
});
